### PR TITLE
Use the best DocumentURI for Document.web_uri

### PR DESCRIPTION
**Pull request by seanh**
Denormalize the "best" DocumentURI.uri that we have into the
Document.web_uri field, instead of just taking the first one.

This fixes an issue that, depending on what order the DocumentURIs are
processed in, a rel-shortlink URI (for example) might be used instead of
a self-claim or canonical URI. We want to display the best available URI
for a document, a self-claim or canonical one if possible, to users.

The first time a given document is annotated consider all of the
DocumentURIs that we've generated from the document metadata sent by the
client and:

- Use a self-claim URI if there is one
- Use a rel-canonical URI if there's no self-claim
- Use any other http(s) URI if there's no self-claim or canonical
- Set web_uri to None if there are not http(s) URIs for the document

I'll need to send a follow-up PR to update the existing web_uris in the production db.

### CODE REVIEW

File: tests/memex/models/document_test.py [hunk 1]
```diff
@@ -833,6 +893,43 @@ def test_it_saves_all_the_document_uris(self,
...
 
+    def test_it_sets_document_web_uri(self,
+                                      annotation,
+                                      best_document_uri_from,
+                                      Document,
+                                      factories,
+                                      session):
+        document_ = mock.Mock(web_uri=None)
+        Document.find_or_create_by_uris.return_value.count.return_value = 1
+        Document.find_or_create_by_uris.return_value.first.return_value = document_
```
Comment (seanh) [inline tests/memex/models/document_test.py]: This mocking is ugly but it's how the existing tests in this class work

File: src/memex/models/document.py [hunk 2]
```diff
@@ -409,6 +405,34 @@ def merge_documents(session, documents, updated=None):
     return master
 
 
+def best_document_uri_from(document_uris):
```
Comment (seanh) [inline src/memex/models/document.py]: I guess just `best_uri_from(docuris)` might be a better name

Comment (nickstenning) [inline src/memex/models/document.py]: I'm not sure `best` is particularly helpful without qualification. Best for what? Identity? Location?

File: src/memex/models/document.py [hunk 3]
```diff
@@ -409,6 +405,34 @@ def merge_documents(session, documents, updated=None):
...
+
+        if urlparse.urlparse(uri).scheme in ['http', 'https']:
+            if type_ == 'self-claim':
+                return uri
+            elif type_ == 'rel-canonical':
+                canonical_uri = canonical_uri or uri
+            else:
+                other_uri = other_uri or uri
+
+    return canonical_uri or other_uri
```
Comment (nickstenning) [inline src/memex/models/document.py]: This seems ok, but personally I find this code a little hard to follow. Having the return statements at two different levels is a bit confusing. Do you think it might be clearer to do something like this?

```python
def _first_url_of_type(uris, type_=None):
    for u in uris:
        if type_ is not None and u.type != type_:
            continue
        if urlparse.urlparse(u.uri).scheme not in ['http', 'https']:
            continue
        return u.uri

return (_first_url_of_type(document_uris, 'self-claim') or 
        _first_url_of_type(document_uris, 'rel-canonical') or
        _first_url_of_type(document_uris))
```

File: src/memex/models/document.py [hunk 4]
```diff
@@ -463,13 +487,19 @@ def update_document_metadata(session,
...
+        document_uris.append(
+            create_or_update_document_uri(
+                session=session,
+                document=document,
+                created=created,
+                updated=updated,
+                **document_uri_dict)
+        )
+
+    if not document.web_uri:
```
Comment (nickstenning) [inline src/memex/models/document.py]: I know this is consistent with the previous code, but shouldn't this denormalised field be updated every time the set of URIs is updated?

File: src/memex/models/document.py (hunk 3 above)
Comment (seanh) [inline src/memex/models/document.py]: I actually rejected this approach because it loops over the uris (up to) 3 times. But it's not likely to be a long list of uris so that's not likely to matter, so if you think it's clearer then sure I can change it

Comment (nickstenning) [inline src/memex/models/document.py]: Why is looping over the URIs a problem? This only happens on write, right?

File: src/memex/models/document.py [hunk 5]
```diff
@@ -281,15 +281,11 @@ def create_or_update_document_uri(session,
...
-    if not document.web_uri:
-        parsed = urlparse.urlparse(uri)
-        if parsed.scheme in ['http', 'https']:
-            document.web_uri = uri
-
     try:
         session.flush()
     except sa.exc.IntegrityError:
         raise ConcurrentUpdateError('concurrent document uri updates')
+    return docuri
```
Comment (seanh) [inline src/memex/models/document.py]: Hmm, yeah I don't think so, I just didn't think of `document.document_uris`, I'll give it a try

File: src/memex/models/document.py (hunk 4 above)
Comment (seanh) [inline src/memex/models/document.py]: You mean that (effectively) web_uri would be updated every time a new annotation of this document is created (or updated probably)? I guess the downside would be potential confusion caused by document domain names changing (e.g. on activity pages), but overall I think we probably should update them yes. We have [the same problem with document titles](https://github.com/hypothesis/product-backlog/issues/127) as well.

I think this is a separate issue but it should be a very simple change to make so I can tag it on to this pr if you want..

File: src/memex/models/document.py (hunk 3 above)
Comment (seanh) [inline src/memex/models/document.py]: Yep, it only happens on write

**Comment by seanh** [comment]
Gonna go for lunch now. I've made all the requested changes, but I think the tests can be deduplicated so will do that after lunch.

**Comment by seanh** [comment]
Ok, all done @nickstenning

**Comment by seanh** [comment]
Crumbs, actually the commit history got a bit messed up, hold on a sec...

**Comment by seanh** [comment]
Ok fixed
